File: archived-examples/mpi-job2/pi-mpi.py
"""
Inefficient calculation of Pi using random numbers and mpi4py

Note that this will only run 100 million iterations per task,
as opposed to 1 billion in the C++ version. The pure-python
calculation is considerably slower.

This uses a trivial numpy array of one 64-bit integer to
communicate between processes as (AFAICT) the reduce operation
is not supported for arbitrary python objects like the standard
arbitrary size integer.
"""
import math
import random

import numpy
from mpi4py import MPI


comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

random.seed(42 + rank)
local_hits = numpy.array([0], dtype=numpy.int64)
global_hits = numpy.array([0], dtype=numpy.int64)

iters = 100_000_000

for idx in range(iters):
    x = random.random()
    y = random.random()
    if x**2 + y**2 < 1.0:
        local_hits[0] += 1

comm.Reduce(local_hits, global_hits, MPI.SUM, 0)

if rank == 0:
   print("Global Attempts: {}".format(iters * size))
   print("Global Hits: {}".format(global_hits[0]))
   pi = 4.0 * global_hits[0] / (iters * size)
   print("Calculated Pi: {}".format(pi))
   print("Ratio calc/actual: {}".format(pi / math.pi))

File: archived-examples/mpi-job2/pi-mpi.cpp
#include "mpi.h"
#include <iostream>
#include <random>
#include <stdint.h>

using namespace std;

int main(int argc, char **argv)
{
    int myid, numprocs;
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &numprocs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myid);

    mt19937 generator(42 + myid);
    uniform_real_distribution<double> dis(0.0, 1.0);
    double x; double y;

    int64_t local_hits = 0LL;
    int64_t global_hits = 0LL;
    int64_t iters = 1000000000LL; // 1 billion

    for (int64_t i = 0LL; i<iters; i++) {
        x = dis(generator);
        y = dis(generator);
        if (x*x + y*y < 1.0) local_hits ++;
    }

    MPI_Reduce(&local_hits, &global_hits, 1, MPI_LONG, MPI_SUM, 0, MPI_COMM_WORLD);
    if(0 == myid) {
        cout.precision(17);
        cout << "Global attempts: " << iters * numprocs << endl;
        cout << "Global hits: " << global_hits << endl;
        double pi = 4.0 * (double) global_hits / (double) (iters * numprocs);
        cout << "Calculated pi: " << pi << endl;
        cout << "Ratio calc/actual: " << pi / 3.14159265358979323846 << endl;
    }
    return 0;
}
